rand_distr/tests/cdf.rs

```diff
@@ -195,3 +565,433 @@ fn binomial() {
+    if x <= 1.0 || x <= a {
+        let mut r2 = a;
+        let mut c2 = 1.0;
+        let mut ans2 = 1.0;
+        loop {
+            r2 += 1.0;
+            c2 *= x / r2;
+            ans2 += c2;
+
+            if c2 / ans2 <= eps {
+                break;
+            }
+        }
+        return ax.exp() * ans2 / a;
+    }
+
+    let mut y = 1.0 - a;
+    let mut z = x + y + 1.0;
+    let mut c = 0;
+
+    let mut p3 = 1.0;
+    let mut q3 = x;
+    let mut p2 = x + 1.0;
+    let mut q2 = z * x;
+    let mut ans = p2 / q2;
+
+    loop {
+        y += 1.0;
+        z += 2.0;
+        c += 1;
+        let yc = y * f64::from(c);
+
+        let p = p2 * z - p3 * yc;
+        let q = q2 * z - q3 * yc;
+
+        p3 = p2;
+        p2 = p;
+        q3 = q2;
+        q2 = q;
+
+        if p.abs() > big {
+            p3 *= big_inv;
+            p2 *= big_inv;
+            q3 *= big_inv;
+            q2 *= big_inv;
+        }
+
+        if q != 0.0 {
+            let nextans = p / q;
+            let error = ((ans - nextans) / nextans).abs();
+            ans = nextans;
+
+            if error <= eps {
+                break;
+            }
+        }
+    }
+    1.0 - ax.exp() * ans
+}
+
+fn gen_harmonic(n: u64, m: f64) -> f64 {
+    match n {
+        0 => 1.0,
+        _ => (0..n).fold(0.0, |acc, x| acc + (x as f64 + 1.0).powf(-m)),
+    }
+}
+
+/// https://docs.rs/spfunc/latest/src/spfunc/zeta.rs.html#20-43
+fn zeta_func(s: f64) -> f64 {
+    if s == 1.0 {
+        return f64::INFINITY;
+    }
+
+    let mut akn = vec![1.0];
+    let mut two_pow = 0.5;
+    let head = 1.0 / (1.0 - 2.0.powf(1.0 - s));
+    let mut tail_prev = 0.0;
+    let mut tail = two_pow * akn[0];
+
+    while (tail - tail_prev).abs() >= f64::EPSILON {
+        update_akn(&mut akn, s);
+        two_pow /= 2.0;
+        tail_prev = tail;
+        tail += two_pow * akn.iter().sum::<f64>();
+    }
+
+    head * tail
+}
+
+fn update_akn(akn: &mut Vec<f64>, s: f64) {
+    let n = akn.len() - 1;
+
+    let n1 = n as f64 + 1.0;
+
+    akn.iter_mut().enumerate().for_each(|(k, a)| {
+        let num = n1;
+        let den = n + 1 - k;
+        *a *= num / den as f64;
+    });
+    let p1 = -1.0 / n1;
+    let p2 = (n1 / (n1 + 1.0)).powf(s);
+
+    akn.push(p1 * p2 * akn[n]);
+}
+
+/// [1] FAST AND ACCURATE CALCULATION OF OWEN’S T-FUNCTION
+///     by Mike Patefield and David Tandy
+///     http://www.jstatsoft.org/v05/i05/paper
+
+fn owen_t(h: f64, a: f64) -> f64 {
+    let absh = h.abs();
+    let absa = a.abs();
+    let ah = absa * absh;
+
+    let mut t;
+    if absa <= 1.0 {
+        t = tf(absh, absa, ah);
+    } else if absh <= 0.67 {
+        t = 0.25 - znorm1(absh) * znorm1(ah) - tf(ah, 1.0 / absa, absh);
+    } else {
+        let normh = znorm2(absh);
+        let normah = znorm2(ah);
+        t = 0.5 * (normh + normah) - normh * normah - tf(ah, 1.0 / absa, absh);
+    }
+
+    if a < 0.0 {
+        t = -t;
+    }
+
+    t
+}
+
+fn tf(h: f64, a: f64, ah: f64) -> f64 {
```

Comment:
what is this function? Where did the code come from?

Reply:
The above code is mentioned in the [paper](http://www.jstatsoft.org/v05/i05/paper) cited in the comment of the `owen_t` function: 

> TF(H,A,AH) (H ≥ 0, 0 ≤ A ≤ 1). This in turn selects the appropriate method (T1,…, T6) for computing T(H,A) based on the input values of H and A according to the ranges given in Figure 2.

The paper provides Fortran code at the end, and I just translated it into Rust.

Reply:
You could embed this function inside `owen_t` then. Or it would be neater to move the lot to a new file.